src/pixelator/analysis/polarization/__init__.py

```diff
@@ -99,60 +103,113 @@ def polarization_scores_component(
+        W = W / W.sum(axis=0)
+        X = create_node_markers_counts(graph)
 
-    # clr transformation
-    if normalization == "clr":
-        counts_df = clr_transformation(df=counts_df, non_negative=True, axis=0)
+        # remove markers with zero variance
+        X = X.loc[:, (X != 0).any(axis=0) & (X.nunique() > 1)]
 
-    # compute the spatial weights matrix (w) from the graph
-    w = WSP(graph.get_adjacency_sparse()).to_W(silence_warnings=True)
+        # Calculate normalization factor
+        C = N / W.sum()
 
-    # compute polarization statistics for each marker using the spatial weights (w)
-    # and the markers counts distribution (y) (Morans I autocorrelation)
-    def data():
-        for m in counts_df.columns:
-            yield morans_autocorr(w, counts_df[m], permutations)
+        X_perm = [
+            clr_transformation(x) if normalization == "clr" else x
+            for x in permutations(X, n=n_permutations, random_seed=random_seed)
+        ]
+
+        if normalization == "clr":
+            X = clr_transformation(X, non_negative=True, axis=0)
 
-    if permutations > 0:
-        df = pd.DataFrame(
-            ((m.I, m.p_rand, m.z_rand, m.p_sim, m.z_sim) for m in data()),
+        _compute_morans_i_per_marker = partial(
+            _compute_morans_i,
+            marker_count_matrix=X,
+            marker_count_matrix_permuted=X_perm,
+            weight_matrix=W,
+            normalization_factor=C,
+        )
+        markers = X.columns
+
+        results = pd.DataFrame(
+            map(_compute_morans_i_per_marker, markers),
             columns=[
+                "marker",
                 "morans_i",
-                "morans_p_value",
                 "morans_z",
-                "morans_p_value_sim",
-                "morans_z_sim",
-            ],
-        ).fillna(0)
-    else:
-        df = pd.DataFrame(
-            ((m.I, m.p_rand, m.z_rand) for m in data()),
-            columns=[
-                "morans_i",
+                "perm_mean",
+                "perm_std",
                 "morans_p_value",
-                "morans_z",
             ],
-        ).fillna(0)
+        )
+        return results
+
+    start_time = time.perf_counter()
+    logger.info("Computing Moran's I for component: %s", component_id)
+
+    results = morans(graph, n_permutations)
+    results["component"] = component_id
+
+    run_time = time.perf_counter() - start_time
+    logger.info(
+        "Finished computing Moran's I for component: %s in %.2fs",
+        component_id,
+        run_time,
+    )
+    results = results.drop(columns=["perm_mean", "perm_std"])
+    return results
+
 
-    df["marker"] = counts_df.columns.tolist()
-    df["component"] = component_id
+def polarization_scores_component_df(
+    component_id: str,
+    component_df: pd.DataFrame,
+    use_full_bipartite: bool,
+    normalization: PolarizationNormalizationTypes,
+    n_permutations: int,
+    random_seed: int | None,
```

Comment:
Good point. I have just missed adding them here. I have made them the same as on the graph function now. And I also changed the default of `n_permutations` from `0` to `50`, so that we have the same defaults in the internal APIs as what we put on the cli.